hal/source/gpio_device.cpp

```diff
@@ -0,0 +1,162 @@
+	nrt_fpga_device_.Set(NRTFPGADevice::kDOData, out_data);
+	SetInitialized();
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, vector<uint8_t>& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<uint8_t> get_data;
+	result.resize(2);
+	if (key == kDIPDataName || key == kDIHDataName)
+	{
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDIData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		result[0] = get_data[0];
+		result[1] = get_data[1];
+	}
+	else if (key == kDOPDataName || key == kDOHDataName)
+	{
+		if (nrt_fpga_device_.Get(NRTFPGADevice::kDOData, get_data) != kNoError)
+			return kInvalidArgument;
+
+		result[0] = get_data[0];
+		result[1] = get_data[1];
+	}
+	else
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Get(const string& key, unsigned int& result) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<string> destination;
+	vector<string> split_data;
+	vector<string>::iterator it;
+	boost::split(destination, key, boost::is_any_of( "." ), \
+			boost::token_compress_on);
+	for (it = destination.begin(); it != destination.end(); ++it)
+		split_data.push_back(*it);
+
+	if (split_data.size() != 2)
+		return kInvalidArgument;
+
+	vector<uint8_t> get_data;
+	int position = stoi(split_data[1]);
+
+	if (Get(split_data[0], get_data) != kNoError)
+		return kInvalidArgument;
+
+	result = (get_data[0] | ((get_data[1] << 8) & 0xff00)) >> (position - 1)& 1;
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Set(const string& key, \
+		const vector<uint8_t>& value) {
+	if (value.size() < sizeof(uint16_t))
+		return kInvalidArgument;
+	vector<uint8_t> get_data;
+	vector<uint8_t> set_data(4);
+	if (key == kDOPDataName)
+	{
+		if (Get(kDOHDataName, get_data) != kNoError)
+			return kInvalidArgument;
+
+		set_data[0] = value[0];
+		set_data[1] = value[1];
+		set_data[2] = get_data[0];
+		set_data[3] = get_data[1];
+	}
+	else if (key == kDOHDataName)
+	{
+		if (Get(kDOPDataName, get_data) != kNoError)
+			return kInvalidArgument;
+
+		set_data[0] = get_data[0];
+		set_data[1] = get_data[1];
+		set_data[2] = value[0];
+		set_data[3] = value[1];
+	}
+	else
+		return kInvalidArgument;
+
+	if (nrt_fpga_device_.Set(NRTFPGADevice::kDOData, set_data) != kNoError)
+		return kInvalidArgument;
+
+	return kNoError;
+}
+
+error_condition GPIODevice::Set(const string& key, const unsigned int value) {
+	if (key.empty())
+		return kInvalidArgument;
+	vector<string> destination;
+	vector<string> split_data;
+	vector<string>::iterator it;
+	boost::split(destination, key, boost::is_any_of( "." ), \
+			boost::token_compress_on);
+	for (it = destination.begin(); it != destination.end(); ++it)
+		split_data.push_back(*it);
+
+	if (split_data.size() != 2)
+		return kInvalidArgument;
+
+	vector<uint8_t> set_data(2);
+	vector<uint8_t> get_data;
+
+	int position = stoi(split_data[1]);
+	if (split_data[0] == kDOPDataName || split_data[0] == kDOHDataName)
```

Comment:
IMHO here should do the early return, e.g. 
```
if (split_data[0] != kDOPDataName && split_data[0] != kDOHDataName)
	return kInvalidArgument;
```
We'll have less indentation if we are using the early return rule.